tests/test_request.py

```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+
+import pytest
+
+from resolver.request import Request
+
+
+class TestRequestResolver:
+    resolver = Request(argument=None)
+
+    @pytest.mark.parametrize(
+        "arg, expected",
+        [
+            (None, ValueError),
+            ("", ValueError),
+            ("invalid", ValueError),
+            ("foo://acme.org/invalid", ValueError),
+        ],
+    )
+    def test_resolving_with_invalid_args(self, arg, expected):
+        with pytest.raises(ValueError):
+            self.resolver.argument = arg
+            self.resolver.resolve()
+
+    def test_resolving_with_valid_arg(self):
+        endpoint = "https://dog.ceo/api/breed/hound/list"
+        self.resolver.argument = endpoint
+        response = self.resolver.resolve()
+        assert (
+            response
+            == '{"message":["afghan","basset","blood","english","ibizan","plott","walker"],'
+            '"status":"success"}'
+        )
```

Comment:
Our tests shouldn't be making real API requests. A better approach I've found that works well is to take your `request_func` as a keyword-only argument to `__init__` and then you can inject it in your unit tests.

Like this: 

```python
class Request(Resolver):
    def __init__(self, argument, stack=None, *, request_func=requests.get):
        super().__init__(argument, stack)
        self.request_get = request_func

    def resolve(self):
        result = self.request_get(self.argument)
        # and so on
```

When you do this, then your tests can explicitly inject a Mock request.get function. If you use `unittest.mock.create_autospec` to mock requests.get, it'll even enforce the correct signature on your mock and fail tests if get is being invoked with an improper signature.

Reply:
Another option is to use [requests-mock](https://requests-mock.readthedocs.io/en/latest/mocker.html) with the [pytest fixture](https://requests-mock.readthedocs.io/en/latest/pytest.html), but that requires serialized tests because it's not thread-safe, so passing in a `request_func` is probably better.

Reply:
Yeah, I don't like using libraries like `requests-mock` when I can get away with it because they still make your tests effectively integration tests. They're an excuse not to properly invert control and inject your dependencies and tend to lead to way too many concerns about external dependencies bleeding into your tests.

Reply:
done, I just used magic mock to inject fake data.

Reply:
> Would you be able to provide example(s) (in yaml) of what you are thinking of @dboitnot ?

```yaml
cats: !request
  url: http://catbase.com/api/v2/cats
  method: POST
  format: json
  form:
    - key: color
      value: black
```

> What does this mean @dboitnot ? Isn't this resolver using the request library to do that?

I'm thinking that

```yaml
cats: !request 'http://catbase.com/api/v2/cats'
```

would be equivalent to:

```yaml
cats: !request
  url: http://catbase.com/api/v2/cats
  method: GET
```

Which I think is the functionality you currently have.